x-pack/solutions/security/plugins/security_solution/public/flyout/document_details/shared/hooks/use_highlighted_fields_privilege.tsx

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useMemo } from 'react';
+import { i18n } from '@kbn/i18n';
+import { useMlCapabilities } from '../../../../common/components/ml/hooks/use_ml_capabilities';
+import { usePrebuiltRuleCustomizationUpsellingMessage } from '../../../../detection_engine/rule_management/logic/prebuilt_rules/use_prebuilt_rule_customization_upselling_message';
+import {
+  explainLackOfPermission,
+  hasUserCRUDPermission,
+} from '../../../../common/utils/privileges';
+import { hasMlLicense } from '../../../../../common/machine_learning/has_ml_license';
+import { hasMlAdminPermissions } from '../../../../../common/machine_learning/has_ml_admin_permissions';
+import { useUserData } from '../../../../detections/components/user_info';
+import { isMlRule } from '../../../../../common/machine_learning/helpers';
+import type { RuleResponse } from '../../../../../common/api/detection_engine';
+
+interface UseHighlightedFieldsPrivilegeParams {
+  /**
+   * The rule to be edited
+   */
+  rule: RuleResponse;
+  /**
+   * Whether the rule exists
+   */
+  isExistingRule: boolean;
+}
+
+interface UseHighlightedFieldsPrivilegeResult {
+  /**
+   * Whether edit highlighted fields button is disabled
+   */
+  isDisabled: boolean;
+  /**
+   * The tooltip content
+   */
+  tooltipContent: string;
+}
+
+/**
+ * Returns whether the edit highlighted fields button is disabled and the tooltip content
+ */
+export const useHighlightedFieldsPrivilege = ({
+  rule,
+  isExistingRule,
+}: UseHighlightedFieldsPrivilegeParams): UseHighlightedFieldsPrivilegeResult => {
+  const [{ canUserCRUD }] = useUserData();
+  const mlCapabilities = useMlCapabilities();
+  const hasMlPermissions = hasMlLicense(mlCapabilities) && hasMlAdminPermissions(mlCapabilities);
+
+  const isEditRuleDisabled =
+    !isExistingRule ||
+    !hasUserCRUDPermission(canUserCRUD) ||
+    (isMlRule(rule?.type) && !hasMlPermissions);
+
+  const upsellingMessage = usePrebuiltRuleCustomizationUpsellingMessage(
+    'prebuilt_rule_customization'
+  );
+
+  const isDisabled = isEditRuleDisabled || (Boolean(upsellingMessage) && rule.immutable);
+
+  const tooltipContent = useMemo(() => {
+    const explanation = explainLackOfPermission(
+      rule,
+      hasMlPermissions,
+      true, // default true because we don't need the message for lack of action privileges
+      canUserCRUD
+    );
+
+    if (isEditRuleDisabled && explanation) {
+      return explanation;
+    }
```

Comment:
Hmmmm 🤔 I think something does not add up here...

If we have the `rule` prop defined, but the `isExistingRule` set to `false` because it was removed, then could we get into a situation where `isEditRuleDisabled` could be `true`, but we wouldn't have any `explanation`?

Reply:
Also I just tried to delete a rule after having generated alerts, then go to the flyout, and the `useRuleWithFallback` hook returned me
```typescript
{
  rule: null;
  existingRule: false;
}
```

I guess this is just the case on local maybe, and there could be scenarios where while we have `existingRule: false` we still are able to retrieve the rules infos...

Reply:
I guess what I'm trying to say is, is there a way to simplify this whole logic but not necessarily needing both `rule` and `existingRule` props in our top level context... But maybe not 😆

Reply:
Thanks! I have since uncovered a bug: for some reason when fetching the rule in context, it does not update if the rule is deleted. I have since moved back the fetch rule logic in highlighted fields, hence the rule related contexts are removed. 

Once that bug is fixed, the scenario you mentioned do surface itself, I added another if logic to account for the tooltip

Reply:
Very nice, I like the loading then the disabled button with tooltip. Awesome!